lxd/api_cluster.go

```diff
@@ -2933,3 +2977,705 @@ func restoreClusterMember(d *Daemon, r *http.Request) response.Response {
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	requestor := request.CreateRequestor(r)
+	d.State().Events.SendLifecycle(project.Default, lifecycle.ClusterGroupUpdated.Event(name, requestor, log.Ctx{"description": req.Description, "members": req.Members}))
+
+	return response.EmptySyncResponse
+}
+
+// swagger:operation PATCH /1.0/cluster/groups/{name} cluster-groups cluster_group_patch
+//
+// Update the cluster group
+//
+// Updates the cluster group configuration.
+//
+// ---
+// consumes:
+//   - application/json
+// produces:
+//   - application/json
+// parameters:
+//   - in: body
+//     name: cluster group
+//     description: cluster group configuration
+//     required: true
+//     schema:
+//       $ref: "#/definitions/ClusterGroupPut"
+// responses:
+//   "200":
+//     $ref: "#/responses/EmptySyncResponse"
+//   "400":
+//     $ref: "#/responses/BadRequest"
+//   "403":
+//     $ref: "#/responses/Forbidden"
+//   "412":
+//     $ref: "#/responses/PreconditionFailed"
+//   "500":
+//     $ref: "#/responses/InternalServerError"
+func clusterGroupPatch(d *Daemon, r *http.Request) response.Response {
+	name := mux.Vars(r)["name"]
+
+	clustered, err := cluster.Enabled(d.db)
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	if !clustered {
+		return response.BadRequest(fmt.Errorf("This server is not clustered"))
+	}
+
+	var clusterGroup *api.ClusterGroup
+	var dbClusterGroup *db.ClusterGroup
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		dbClusterGroup, err = tx.GetClusterGroup(name)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	clusterGroup, err = dbClusterGroup.ToAPI()
+	if err != nil {
+		return response.SmartError(err)
+	}
+
+	req := clusterGroup.Writable()
+
+	// Validate the ETag.
+	etag := []interface{}{clusterGroup.Description, clusterGroup.Members}
+	err = util.EtagCheck(r, etag)
+	if err != nil {
+		return response.PreconditionFailed(err)
+	}
+
+	// Parse the request.
+	err = json.NewDecoder(r.Body).Decode(&req)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	if req.Members == nil {
+		req.Members = clusterGroup.Members
+	}
+
+	err = d.cluster.Transaction(func(tx *db.ClusterTx) error {
+		obj := db.ClusterGroup{
+			Name:        dbClusterGroup.Name,
+			Description: req.Description,
+		}
+
+		err = tx.UpdateClusterGroup(name, obj)
+		if err != nil {
+			return err
+		}
+
+		members, err := tx.GetClusterGroupNodes(name)
+		if err != nil {
+			return err
+		}
+
+		// skipMembers is a list of members which already belong to the group.
+		skipMembers := []string{}
+
+		for _, oldMember := range members {
+			if !shared.StringInSlice(oldMember, req.Members) {
+				// Get all cluster groups this member belongs to.
+				groups, err := tx.GetClusterGroupsWithNode(oldMember)
+				if err != nil {
+					return err
+				}
+
+				// Note that members who only belong to this group will not be removed from it.
```

Comment:
Should we return an error in this scenario rather than silently only partially completing the request?